jetty-websocket/websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/DemandingFlusher.java

```diff
@@ -0,0 +1,181 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2021 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.websocket.core.internal;
+
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.LongConsumer;
+
+import org.eclipse.jetty.util.Callback;
+import org.eclipse.jetty.util.CountingCallback;
+import org.eclipse.jetty.util.IteratingCallback;
+import org.eclipse.jetty.websocket.core.Extension;
+import org.eclipse.jetty.websocket.core.Frame;
+
+/**
+ * <p>This flusher can be used to mutated and fragment {@link Frame}s and forwarded them on towards the application using the
+ * {@link BiConsumer} provided in the constructor. This can split a single incoming frame into n {@link Frame}s which are
+ * passed on to the {@link BiConsumer} one at a time.</p>
+ *
+ * <p>The asynchronous operation performed by this {@link IteratingCallback} is demanding from upper layer after which
+ * {@link #onFrame(Frame, Callback)} will called with the new content.</p>
+ *
+ * <p>This flusher relies on the interception of demand, and because of this it can only be used in an {@link Extension} which
+ * implements the {@link DemandChain} interface. The methods of {@link DemandChain} from the {@link Extension}
+ * must be forwarded to this flusher.</p>
+ */
+public abstract class DemandingFlusher extends IteratingCallback implements DemandChain
+{
+    private final BiConsumer<Frame, Callback> _emitFrame;
+    private final AtomicLong _demand = new AtomicLong();
+    private final AtomicReference<Throwable> _failure = new AtomicReference<>();
+    private LongConsumer _nextDemand;
+
+    private Frame _frame;
+    private Callback _callback;
+    private boolean _needContent = true;
+    private boolean _first = true;
+
+    /**
+     * @param emitFrame where frames generated by {@link #handle(Frame, Callback, boolean)} are forwarded.
```

Comment:
```suggestion
     * @param emitFrame The BiConsumer to which generated frames are forwarded, along with a callback to succeed when the frame can be released, by {@link #handle(Frame, Callback, boolean)}.
```

We also need to clarify somewhere that `callback.failed(t)` means.   Perhaps that can be javadoc'd in one place (probably `IncomingFrames`) and then whenever  a like callback is used we can refer to that.   Perhaps also we should make the association between the BiConsumer and `IncomingFrames.onFrame`.... actually why isn't this just `IncomingFrames` rather than `BiConsumer`?